mrc-2312 Add workflow progress page

Nicolas-Dolan opened the pull request:
_Before submitting please review
the [PR checklist](https://mrc-ide.myjetbrains.com/youtrack/articles/RESIDE-A-13/Pull-Request-guidelines)_

Nicolas-Dolan commented on src/app/static/package.json:
```diff
@@ -56,5 +56,8 @@
     "webpack-stream": "^5.2.1"
+  },
+  "dependencies": {
+    "vue-select": "^3.11.2"
```
does this need to be a dev-dependency?

mwoodbri replied:
Yep - see mrc-2308

Nicolas-Dolan commented on src/app/static/src/js/components/runWorkflow/runWorkflowProgress.vue:
```diff
@@ -1,19 +1,191 @@
+            </table>
+        </div>
+        <div class="row justify-content-end mt-3">
+            <button class="button mr-3" type="button" disabled>
+                Clone workflow
+            </button>
+            <button class="btn btn-grey" type="button" disabled>
+                Cancel workflow
+            </button>
```
These are just placeholders for now.

mwoodbri replied:
mrc-2513

Nicolas-Dolan commented on src/app/static/src/tests/components/runWorkflow/runWorkflowProgress.test.ts:
```diff
@@ -18,4 +90,53 @@ describe(`runWorkflowProgress`, () => {
+        await Vue.nextTick()
+        await Vue.nextTick()
+        await Vue.nextTick()
+        await wrapper.setData({workflowRunStatus: workflowStatus1})
```
I was unable to interact with v-select directly with unit tests so have covered the format of the items in the dropdown and the behaviour when an item is selected in the selenium tests instead.

Nicolas-Dolan replied:
I had tried using a full mount but I couldn't get that to work either unfortunately.

Nicolas-Dolan commented on src/app/templates/index.ftl:
```diff
@@ -32,6 +32,10 @@
                 </a>
             </button>
         </#if>
+        <button id="run-workflow" class="ml-5 btn-link btn btn-lg"><a href="${appUrl}/run-workflow">Run workflow
+                <@customSvg />
+            </a>
+        </button>
```
This can be removed if conflicts with other tickets.

mwoodbri replied:
Yes, see mrc-2326. This button should only be made visible when all other workflows work is complete.

EmmaLRussell commented on src/customConfigTests/src/test/kotlin/org/vaccineimpact/orderlyweb/customConfigTests/RunWorkflowTests.kt:
```diff
@@ -33,4 +36,24 @@ class RunWorkflowTests : SeleniumTest()
+        dropdownMenu[0].click()
+        wait.until(ExpectedConditions.presenceOfElementLocated(By.id("workflow-table")))
+        val table = driver.findElement(By.id("workflow-table"))
+        Assertions.assertThat(table.text).contains("Reports")
```
So it's getting as far as showing the table, so the status must be set, but there's nothing rendered in the table, i.e. don't seem to be any reports in the status? Is that because the status endpoint fetches status from orderly, and orderly doesn't know about the workflow - it's just had a row inserted into the OW db table? Might need to actually run the workflow through OrderlyServerAPI?

Nicolas-Dolan replied:
Are there any examples of that having been done before that I can look at?

EmmaLRussell replied:
No I don't think so, but I think it *should* "just work" if you create an [OrderlyServer](https://github.com/vimc/orderly-web/blob/master/src/app/src/main/kotlin/org/vaccineimpact/orderlyweb/OrderlyServerAPI.kt#L49) object and give it an [AppConfig](https://github.com/vimc/orderly-web/blob/master/src/databaseInterface/src/main/kotlin/org/vaccineimpact/orderlyweb/db/AppConfig.kt#L15) object as its parameter, I think it should read in the default config and be able to connect to orderly running in the dependencies.

EmmaLRussell reviewed (changes requested):
This is looking really good so far.

Nicolas-Dolan commented on src/app/static/src/js/components/runWorkflow/runWorkflowProgress.vue:
```diff
@@ -1,19 +1,191 @@
+                    <td v-else>{{ report.name }}</td>
+                    <td :style="{ color: statusColour(report.status) }">
+                        {{ report.status }}
+                    </td>
```
The endpoint does not currently return the date for each report. This needs to be adjusted on orderly server first. The frontend is ready for a date value though, as seen in the line below and unit tests.

EmmaLRussell replied:
Ah sorry, yeah, I was just looking at the date for the workflow

Nicolas-Dolan commented on src/app/static/src/js/components/runWorkflow/runWorkflowProgress.vue:
```diff
@@ -1,19 +1,206 @@
+                            >{{ report.name }}</a
+                        >
+                    </td>
+                    <td v-else-if="report.status === 'error'">
+                        <a
+                            :href="reportLogsHref(report.name)"
```
Are only error messages going to the logs page? Or should missing, impossible, etc go there as well?

EmmaLRussell replied:
The logs page shows all the logging for any report, whatever its status - as well as showing the status itself.

Nicolas-Dolan commented on src/app/static/src/js/components/runWorkflow/runWorkflowProgress.vue:
```diff
@@ -1,19 +1,206 @@
+            return buildFullUrl(url);
+        },
+        reportLogsHref(name) {
+            const url = `/run-report?report-name=${name}`; // full implementation into report logs will be covered in a future ticket
```
If the name contains spaces, will that be automatically fused when it is made into a url or do I have to do that here?

mwoodbri replied:
The name [shouldn't include spaces](https://www.vaccineimpact.org/orderly/articles/orderly.html#creating-a-report) and I don't think it will include any characters needing URL-encoding so I think you're fine here

Nicolas-Dolan replied:
My memory is failing me as this was so long ago but I believe I meant that run reports would handle the url params to bring up the appropriate report logs. This is alluded to in the mock ups.

EmmaLRussell reviewed (commented):
Thanks, that's addressed most of my comments, have resolved those. You're still working on tests though, aren't you, and Buildlkte is failing.

mwoodbri reviewed (changes requested):
A few issues to be resolved.

Please also remove `package-lock.json` from this PR as it's not accompanied by a `package.json` change.

Incidentally, I noticed a server bug when testing this: mrc-2475. When it occurs it is exhibited by clicking on a completed report and getting a "report not found" error.

Nicolas-Dolan commented:
I've removed the selenium test that I added as the whole suite have been randomly failing locally even when they pass in buildkite, making them very difficult to debug. If anyone has any guidance on making them more stable so that I can test them locally, I'd be interested to hear.

mwoodbri commented on src/app/static/src/js/components/runWorkflow/runWorkflowProgress.vue:
```diff
@@ -1,19 +1,208 @@
+        v-if="workflowRunSummaries"
+    >
+        <div class="row mb-3">
+            <label for="workflows" class="form-label col">Workflow</label>
```
`for` needs to identify an element by ID, so you can either change the `id` of the <v-select> to `workflows` or (less nice) do `<label for="v-select" ...>`

mwoodbri commented on src/app/static/src/js/components/runWorkflow/runWorkflowProgress.vue:
```diff
@@ -1,19 +1,208 @@
+    WorkflowRunStatus,
+} from "../../utils/types";
+import { buildFullUrl } from "../../utils/api";
+import { session } from "./../../utils/session.js";
```
```suggestion
import { session } from "../../utils/session.js";
```

mwoodbri commented on src/app/static/src/js/components/runWorkflow/runWorkflowProgress.vue:
```diff
@@ -1,19 +1,208 @@
+                            >{{ report.name }}</a
+                        >
+                    </td>
+                    <td v-else-if="report.status === 'error'" class="p-2">
+                        <a
+                            :href="reportLogsHref(report.name)"
+                            @click="setReportLogsTab"
+                            >{{ report.name }}</a
+                        >
+                    </td>
```
I don't think there's value in hyperlinking this, given it doesn't point at anything. As noted in the mockups, exactly how we display failure logs is an open question. Logged as mrc-2514.

mwoodbri reviewed (changes requested):
This is all looking very tidy!

A Selenium test is definitely warranted IMO. Invaluable to ensure that the page renders correctly with real server responses, amongst other things. Please restore whatever you had and I'm sure we can collectively solve any outstanding issues.

carbowizzy commented on src/app/static/src/tests/components/runWorkflow/runWorkflowProgress.test.ts:
```diff
@@ -1,15 +1,88 @@
+        setTimeout(() => {
+            expect(wrapper.find("label").text()).toBe("Workflow")
+            expect(wrapper.find("v-select-stub").attributes("placeholder")).toBe("Select workflow or search by name...")
+            // Tests to be reinstated once buttons are integrated with workflow wizard
+            // expect(wrapper.findAll("button").at(0).text()).toBe("Clone workflow")
+            // expect(wrapper.findAll("button").at(1).text()).toBe("Cancel workflow")
```
This comment lines can be removed. Instead check that the buttons are displayed but disabled.

EmmaLRussell replied:
That won't work for the current implementation, the buttons aren't currently rendered, using `v-if="false"` - might be better just to have a test for no buttons here, which will fail and need to be modified when the buttons are added.

carbowizzy reviewed (changes requested):
This is looking really good.

I have not been able to test manually. I have some suggestions below.

EmmaLRussell commented on src/app/static/src/js/components/runWorkflow/runWorkflowProgress.vue:
```diff
@@ -31,34 +31,35 @@
         </div>
-        <div class="row justify-content-end mt-3">
-            <button class="button mr-3" type="button" disabled>
+        <!-- Buttons to be unhidden and made active once workflow wizard is implemented -->
```
Is there a ticket for this? The wizard is pretty much done now, so that should be possible to do.

Nicolas-Dolan replied:
mrc-2513

EmmaLRussell commented on src/app/static/src/js/components/runWorkflow/runWorkflowProgress.vue:
```diff
@@ -1,19 +1,208 @@
+            </table>
+        </div>
+        <!-- Buttons to be unhidden and made active once workflow wizard is implemented -->
+        <div class="row justify-content-end mt-3" v-if="false">
```
To be fair, this is what the mockup looks like, and the rest of the workflow steps deviate from the mockups. Maybe it would be better to bring the wizard steps into line with the mockups so that action buttons are aligned right.

mwoodbri commented on src/app/static/src/js/components/runWorkflow/runWorkflowProgress.vue:
```diff
@@ -1,19 +1,208 @@
+                </v-select>
+            </div>
+        </div>
+        <div class="row" v-if="workflowRunStatus" id="workflow-table">
```
The implementation plan states that this information should be updated automatically by polling the server, as we do for report progress. This should be a reasonably simple addition via a `setInterval()`. Otherwise there's no way to get updates - there's no refresh button and reloading the page deselects the relevant workflow.

EmmaLRussell replied:
Maybe that could be spun out into a separate ticket, it would be really good to get this merged.

mwoodbri replied:
Fine me with: @Nicolas-Dolan please create a ticket if you agree

Nicolas-Dolan replied:
mrc-2516

EmmaLRussell reviewed (commented):
Looking good, I'd definitely vote for reinstating the Selenium test, particularly if it passes reliably on Buildkite. Selenium tests seems reliable for me locally, have you noticed any pattern to the failures? I think some can fail if you're running the app as well and have done stuff in it, as it can find unexpected state. (Hm, or I might be mixing that up with the integration tests actually..)

Nicolas-Dolan commented on src/customConfigTests/src/test/kotlin/org/vaccineimpact/orderlyweb/customConfigTests/RunWorkflowTests.kt:
```diff
@@ -221,4 +223,24 @@ class RunWorkflowTests : SeleniumTest()
+        dropdownMenu[0].click()
+        wait.until(ExpectedConditions.presenceOfElementLocated(By.id("workflow-table")))
+        val table = driver.findElement(By.id("workflow-table"))
+        assertThat(table.text).contains("Reports")
```
@EmmaLRussell this should contain the reports in the table, not just the header. I cannot get them to render. I tried inserting a report with the helper into the database but it isn't showing up. Do you have any suggestions?

EmmaLRussell replied:
I think the problem is that, although we've told our database tables about the workflow in the setup method, orderly.server doesn't know about it, so we're not going to get a sensible response from the status endpoint. So we need to run a workflow for real - can't do this through the UI until the Submit ticket is done, but for now we can invoke some javascript with Mark's cool fetch script. 

This works for me, I'll leave you to add any further tests of the table you want

EmmaLRussell commented on src/customConfigTests/src/test/kotlin/org/vaccineimpact/orderlyweb/customConfigTests/RunWorkflowTests.kt:
```diff
@@ -221,4 +223,24 @@ class RunWorkflowTests : SeleniumTest()
+    @Test
+    fun `can select workflow progress tab and selecting a workflow option generates reports table`()
+    {
+        val link = driver.findElement(By.id("workflow-progress-link"))
+        assertThat(link.text).isEqualTo("Workflow progress")
+        link.click()
+        wait.until(ExpectedConditions.presenceOfElementLocated(By.id("workflow-progress-tab")))
+        val vSelectInput = driver.findElement(By.tagName("input"))
+        vSelectInput.sendKeys("workf")
+
+        val vSelect = driver.findElement(By.id("workflows"))
+        val dropdownMenu = vSelect.findElements(By.tagName("li"))
+        assertThat(dropdownMenu[0].text).contains("workflow1\n" +
+            "Tue Jun 15 2021, 14:50")
+        dropdownMenu[0].click()
+        wait.until(ExpectedConditions.presenceOfElementLocated(By.id("workflow-table")))
+        val table = driver.findElement(By.id("workflow-table"))
```
```suggestion
       //NB This should be replaced with running a workflow through the UI once workflow submit is implemented
       val jse = driver as JavascriptExecutor
       jse.executeScript("""fetch("${RequestHelper.webBaseUrl}/workflow", {"method": "POST", "body": "{\"name\":\"My    workflow\",\"reports\":[{\"name\":\"minimal\"},{\"name\":\"global\"}],\"changelog\":{\"message\":\"message1\",\"type\":\"internal\"}}"})""")

      val link = driver.findElement(By.id("workflow-progress-link"))
      assertThat(link.text).isEqualTo("Workflow progress")
      link.click()
      wait.until(ExpectedConditions.presenceOfElementLocated(By.id("workflow-progress-tab")))
      val vSelectInput = driver.findElement(By.tagName("input"))
      vSelectInput.sendKeys("workf")

      val vSelect = driver.findElement(By.id("workflows"))
      val dropdownMenu = vSelect.findElements(By.tagName("li"))
      assertThat(dropdownMenu[0].text).contains("My workflow")
      dropdownMenu[0].click()
      wait.until(ExpectedConditions.presenceOfElementLocated(By.id("workflow-table")))
      val table = driver.findElement(By.id("workflow-table"))
      assertThat(table.text).contains("Reports")
      val rows = driver.findElements(By.cssSelector("#workflow-table tr"))
      assertThat(rows.count()).isEqualTo(2)
```

Nicolas-Dolan commented on src/customConfigTests/src/test/kotlin/org/vaccineimpact/orderlyweb/customConfigTests/RunWorkflowTests.kt:
```diff
@@ -181,7 +184,7 @@ class RunWorkflowTests : SeleniumTest()
         wait.until(ExpectedConditions.presenceOfElementLocated(By.id("git-branch")))
 
-        assertThat(driver.findElements(By.cssSelector("#git-commit option")).count()).isEqualTo(1)
+        assertThat(driver.findElements(By.cssSelector("#git-commit option")).count()).isEqualTo(2)
```
@EmmaLRussell I had to change this to get the test suite to pass. Do you think this is problematic or a legitimate change?

EmmaLRussell replied:
This is kind of a 'one strike and you're out' test - it changes the state of the orderly container since refreshing git causes the fake git system to add a new commit (I *think* it will only do this up to a maximum of two) and that's what the test checks for. It should be passing on Buildkite, but it will fail locally if you don't restart the deps between test runs. I've just tested it locally, and that was the behaviour I was seeing. I don't think the test should need to change to pass on BK - it's probably best just to ignore all the other tests and concentrate of getting your workflow progress test working locally, and let BK do the rest.

EmmaLRussell commented on src/customConfigTests/src/test/kotlin/org/vaccineimpact/orderlyweb/customConfigTests/RunWorkflowTests.kt:
```diff
@@ -221,4 +224,29 @@ class RunWorkflowTests : SeleniumTest()
+    {
+        //NB This should be replaced with running a workflow through the UI once workflow submit is implemented
+        val jse = driver as JavascriptExecutor
+        jse.executeScript("""fetch("${RequestHelper.webBaseUrl}/workflow", {"method": "POST", "body": "{\"name\":\"workflow1\",\"reports\":[{\"name\":\"minimal\"},{\"name\":\"global\"}],\"changelog\":{\"message\":\"message1\",\"type\":\"internal\"}}"})""")
```
```suggestion
        jse.executeAsyncScript(
        "var callback = arguments[arguments.length - 1];" +
        """await fetch("${RequestHelper.webBaseUrl}/workflow", {"method": "POST", "body": "{\"name\":\"My    workflow\",\"reports\":[{\"name\":\"minimal\"},{\"name\":\"global\"}],\"changelog\":{\"message\":\"message1\",\"type\":\"internal\"}}"});""" +
        "callback();"
        )
```

mwoodbri commented on src/app/static/src/tests/components/runWorkflow/runWorkflowProgress.test.ts:
```diff
@@ -1,14 +1,121 @@
 import {shallowMount} from "@vue/test-utils";
 import runWorkflowProgress from '../../../js/components/runWorkflow/runWorkflowProgress.vue'
+import {mockAxios} from "../../mockAxios";
+import Vue from "vue";
```
Is this used anywhere?

EmmaLRussell commented on src/customConfigTests/src/test/kotlin/org/vaccineimpact/orderlyweb/customConfigTests/RunWorkflowTests.kt:
```diff
@@ -181,7 +184,7 @@ class RunWorkflowTests : SeleniumTest()
         wait.until(ExpectedConditions.presenceOfElementLocated(By.id("git-branch")))
 
-        assertThat(driver.findElements(By.cssSelector("#git-commit option")).count()).isEqualTo(1)
+        assertThat(driver.findElements(By.cssSelector("#git-commit option")).count()).isIn(listOf(1, 2))
```
This did seem to be necessary for Buiildkite, for slightly mysterious reasons (Rob did think it might be related to running a workflow). Agree, a comment would be good, and maybe an additional check for no error message.

EmmaLRussell commented on src/customConfigTests/src/test/kotlin/org/vaccineimpact/orderlyweb/customConfigTests/RunWorkflowTests.kt:
```diff
@@ -221,4 +224,33 @@ class RunWorkflowTests : SeleniumTest()
+    {
+        //NB This should be replaced with running a workflow through the UI once workflow submit is implemented
+        val jse = driver as JavascriptExecutor
+        jse.executeAsyncScript(
+            "var callback = arguments[arguments.length - 1];" +
+            """await fetch("${RequestHelper.webBaseUrl}/workflow", {"method": "POST", "body": "{\"name\":\"My    workflow\",\"reports\":[{\"name\":\"minimal\"},{\"name\":\"global\"}],\"changelog\":{\"message\":\"message1\",\"type\":\"internal\"}}"});""" +
+            "callback();"
```
Ah yes, sorry @Nicolas-Dolan for some reason I thought `executeScript` didn't wait for completion, but looks like it does

EmmaLRussell reviewed (commented):
Nothing much to add, but I do think a comment in the git refresh test would be helpful, and also a test for no error, since the test of commit number is now less meaningful.

Nicolas-Dolan commented:
@EmmaLRussell @mwoodbri I accidentally merged this in without thinking. I've reverted the merge but I'm not sure if you can review on here or a new branch is needed.
